src/libraries/System.Linq/src/System/Linq/Shuffle.cs

```diff
@@ -0,0 +1,91 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+
+namespace System.Linq
+{
+    public static partial class Enumerable
+    {
+        /// <summary>Shuffles the order of the elements of a sequence.</summary>
+        /// <typeparam name="TSource">The type of the elements of <paramref name="source"/>.</typeparam>
+        /// <param name="source">A sequence of values to shuffle.</param>
+        /// <returns>A sequence whose elements correspond to those of the input sequence in randomized order.</returns>
+        /// <remarks>Randomization is performed using a non-cryptographically-secure random number generator.</remarks>
+        public static IEnumerable<TSource> Shuffle<TSource>(this IEnumerable<TSource> source)
+        {
+            if (source is null)
+            {
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.source);
+            }
+
+            if (IsEmptyArray(source))
+            {
+                return [];
+            }
+
+            if (source is ShuffleIterator<TSource> shuffleIterator)
```

Comment:
Seems like an obvious optimization, but is it realistically common enough to force another type check?